examples/window-app/nrfconnect/main/WindowCovering.cpp

```diff
@@ -0,0 +1,340 @@
+
+static k_timer sLiftTimer;
+static k_timer sTiltTimer;
+static constexpr uint32_t sMoveTimeoutMs{ 200 };
+
+static constexpr uint32_t FromOneRangeToAnother(uint32_t aInMin, uint32_t aInMax, uint32_t aOutMin, uint32_t aOutMax,
+                                                uint32_t aInput)
+{
+    const auto diffInput  = aInMax - aInMin;
+    const auto diffOutput = aOutMax - aOutMin;
+    if (diffInput > 0)
+    {
+        return static_cast<uint32_t>(aOutMin + static_cast<uint64_t>(aInput - aInMin) * diffOutput / diffInput);
+    }
+    return 0;
+}
+
+WindowCovering::WindowCovering()
+{
+    mLiftLED.Init(LIFT_STATE_LED);
+    mTiltLED.Init(TILT_STATE_LED);
+
+    if (mLiftIndicator.Init(LIFT_PWM_DEVICE, LIFT_PWM_CHANNEL, 0, 255) != 0)
+    {
+        LOG_ERR("Cannot initialize the lift indicator");
+    }
+    if (mTiltIndicator.Init(TILT_PWM_DEVICE, TILT_PWM_CHANNEL, 0, 255) != 0)
+    {
+        LOG_ERR("Cannot initialize the tilt indicator");
+    }
+
+    k_timer_init(&sLiftTimer, MoveTimerTimeoutCallback, nullptr);
+    k_timer_init(&sTiltTimer, MoveTimerTimeoutCallback, nullptr);
+}
+
+void WindowCovering::MoveTimerTimeoutCallback(k_timer * aTimer)
+{
+    if (!aTimer)
+        return;
+
+    if (aTimer == &sLiftTimer)
+    {
+        chip::DeviceLayer::PlatformMgr().ScheduleWork(DriveCurrentLiftPosition);
+    }
+    else if (aTimer == &sTiltTimer)
+    {
+        chip::DeviceLayer::PlatformMgr().ScheduleWork(DriveCurrentTiltPosition);
+    }
+}
+
+void WindowCovering::DriveCurrentLiftPosition(intptr_t)
+{
+    NPercent100ths current{};
+    NPercent100ths target{};
+    NPercent100ths positionToSet{};
+
+    VerifyOrReturn(Attributes::CurrentPositionLiftPercent100ths::Get(Endpoint(), current) == EMBER_ZCL_STATUS_SUCCESS);
+    VerifyOrReturn(Attributes::TargetPositionLiftPercent100ths::Get(Endpoint(), target) == EMBER_ZCL_STATUS_SUCCESS);
+
+    UpdateOperationalStatus(MoveType::LIFT, ComputeOperationalState(target, current));
+
+    positionToSet.SetNonNull(CalculateSingleStep(MoveType::LIFT));
+    LiftPositionSet(Endpoint(), positionToSet);
+
+    // assume single move completed
+    Instance().mInLiftMove = false;
+
+    VerifyOrReturn(Attributes::CurrentPositionLiftPercent100ths::Get(Endpoint(), current) == EMBER_ZCL_STATUS_SUCCESS);
+
+    if (!TargetCompleted(MoveType::LIFT, current, target))
+    {
+        // continue to move
+        StartTimer(MoveType::LIFT, sMoveTimeoutMs);
+    }
+    else
+    {
+        // the OperationalStatus should indicate no-lift-movement after the target is completed
+        UpdateOperationalStatus(MoveType::LIFT, ComputeOperationalState(target, current));
+    }
+}
+
+chip::Percent100ths WindowCovering::CalculateSingleStep(MoveType aMoveType)
+{
+    EmberAfStatus status{};
+    chip::Percent100ths percent100ths{};
+    NPercent100ths current{};
+    OperationalState opState{};
+
+    OperationalStatus opStatus = OperationalStatusGet(Endpoint());
+
+    if (aMoveType == MoveType::LIFT)
```

Comment:
I m would like to consolidate this with WindowCovering common

Reply:
The same as above, can we possibly postpone this consolidation?